actionpack/lib/action_dispatch/journey/formatter.rb

```diff
@@ -67,7 +67,7 @@ def extract_parameterized_parts(route, options, recall, parameterize = nil)
           parameterized_parts = recall.merge(options)
 
           keys_to_keep = route.parts.reverse_each.drop_while { |part|
-            !options.key?(part) || (options[part] || recall[part]).nil?
+            (!options.key?(part) && !value_for_optional_scope?(part, route, recall)) || (options[part] || recall[part]).nil?
```

Comment:
I find this hard to read. Maybe we can refactor, or space out the logic and add some comments?

Reply:
I was able to make the condition a little bit more simple and removed also the auxiliary method `value_for_optional_scope?`.